evil-core.el

```diff
@@ -285,17 +285,39 @@ See also `evil-initial-state'."
         (evil-initial-state major-mode)
         default)))
 
-(defun evil-initial-state (mode &optional default)
+(defun evil--real-function (fun)
+  "Figure out the actual symbol behind a function.
+Returns a different symbol if FUN is an alias, otherwise FUN."
+  (let ((symbol-function (symbol-function fun)))
+    (if (symbolp symbol-function)
+        symbol-function
+      fun)))
+
+(defun evil--derived-mode-p (mode modes)
+  (let ((parent (evil--real-function mode))
+	visited-modes)
+    (while (and parent (not (memq parent modes)))
+      (push parent visited-modes)	; remember already visited parents
+      (setq parent			; get the next parent
+	    (evil--real-function (get parent 'derived-mode-parent)))
+      (when (memq parent visited-modes)
+	(error "Circular list detected. Aborting.")))
+    parent))
+
+(defun evil-initial-state (mode &optional default noinherit)
   "Return the Evil state to use for MODE.
-Returns DEFAULT if no initial state is associated with MODE.
-The initial state for a mode can be set with
+Returns DEFAULT if no initial state is associated with either
+MODE, or any of its parents. If NOINHERIT is not nil, returns
+DEFAULT if no initial state is associated with this mode. The
+initial state for a mode can be set with
 `evil-set-initial-state'."
   (let (state modes)
     (catch 'done
-      (dolist (entry (evil-state-property t :modes) default)
+      (dolist (entry (nreverse (evil-state-property t :modes)) default)
```

Comment:
It's true that I didn't follow that to the end, I just assumed you had a good reason for it. The test passes (on my PC) even without it. Should I take it out?